### ISSUE TITLE
Direct3D 12 support

### ISSUE THREAD

**Issue by oscarbg** [apitrace/apitrace#372]
Hi,
I just see 11.3 support was added recently.. can we expect D3D12 soon? or it's lots of work?
thanks

**Comment by jrfonseca** [comment]
I haven't looked too closely into it.

Tracing shouldn't be too hard, but it's possible that swizlling GPU addresses when retracing will be a tough nut to crack.

I can't give any ETA.  I'll only look into it when I come across a D3D12 app that I want to trace, or as a means to study D3D12 in more detail.

I'll happily accept patches though ;-)

**Comment by oscarbg** [comment]
any updates on this?
if not may be will try to code it myself..

**Comment by jrfonseca** [comment]
I had (and still have) no immediate interest in D3D12.

But given you're volunteering to help, I spent some time today adding enough skeleton code to support tracing D3D12.  It's untested, and I'm sure there's lots of loose ends that need to be tied.  But it should be a much easier to progress from here than start from scratch.

I don't plan to put more time into this myself in the foreseeable future.  (I know I'll need D3D12 at some point, but not at the moment.)  So it's up to you to make progress.

Take a look at d09fd768bfe960f7e03907c9a6b61b0ce63 , see how D3D11 support is implemented, and let me know if you have questions.

**Comment by oscarbg** [comment]
Thanks for the support!
Will post progress here..

**Comment by mtschoen** [comment]
Thoughts on Vulkan?

**Comment by jrfonseca** [comment]
> Thoughts on Vulkan?

@mtschoen yep, see https://github.com/apitrace/apitrace/issues/439#issuecomment-202429352 . Executive summary is: I don't object patches but I don't plan to (ever) work on it myself, and I instead recommend other tools.

**Comment by oscarbg** [comment]
Hi jrfonseca,
I'm stuck with D3D12 more early than expected..
I downloaded Microsoft D3D12 samples DirectX-Graphics-Samples-1.0.160510.1 from Microsoft github
and found two things:
*Can't trace UWP (universal Windows 10 platform/Winrt/Metro) apps: that was expected but posting here in case you know how to inject  dxgitrace.dll to a UWP app..
checked apitrace trace -a dxgi exe launches:
inject -D C:\code\apitrace-d3d12\bin64\Debug\..\wrappers\Debug\dxgitrace.dll -- 
C:\Users\oscarb\Downloads\DirectX-Graphics-Samples-1.0.160510.1\DirectX-Graphics-Samples-1.0.160510.1\Samples\UWP\D3D12Fullscreen\src\bin\x64\Release\D3D12Fullscreen.exe
I get:
inject: error: failed to execute D3D12Fullscreen.exe
as said expected as you can't run also UWP apps from running executable file from command line..
so question is if it's posible to trace even a super simple sample UWP D3D12 app like D3D12Fullscreen..
I ask this because modern D3D12 games like Quantum Break and Forza Apex are D3D12 UWP apps

also should be interesting for you as this should be enable to trace D3D11 apps as even new games in Windows Store like Killer Instint are D3D11 UWP apps..

Another issue is even desktop app doesn't trace D3D12 api calls..
I switch to D3D12Fullscreen Desktop simple (i.e. classical Win32 executable)..
Program is launched correctly and I get a trace but dumping it it has only:
0 CreateDXGIFactory1(riid = IID_IDXGIFactory4, ppFactory = &0x2b52ea0adc0) = S_OK
1 CreateDXGIFactory1(riid = IID_IDXGIFactory4, ppFactory = &0x2b52ea0f1c0) = S_OK
I first searched in injectee.cpp and added d3d12.dll thinking will fix it here below d3d11.dll:
 if (stricmp(szNewDllBaseName, "dxgitrace.dll") == 0) 
..
registerModuleHooks("d3d12.dll", g_hHookModule);

altough I don't know what does this code wasn't expecting that adding that line  executable to trace doesn't even launch..
then I see dxgitrace.def you export D3D12CreateDevice so with depends tool I examined both desktop&UWP app D3D12Fullscreen.exe and I see in executable d3d12.dll file d3d12createdevice isn't mentioned but there is a dependency function which is defined with ordinal 101 which for d3d12.dll is d3d12createdevice searching for strings in executable also shows d3d12createdevice isn't mentioned altough a reference to D3D12SerializeRootSignature is present both in depends and found as string in executable file i.e. D3D12SerializeRootSignature is referenced with function name..
don't know anything about injection but in light of this I modified dxgitrace.def:
D3D12CreateDevice @101
so now depends shows dxgitrace.dll D3D12createdevice funtion has ordinal 101.. anyway injection still fails..
can you help fixing at least injection of D3D12createdevice for desktop apps.. also support for injection of uwp apps would be nice..
also thinking if would be posible to generate a d3d12.dll as way of tracing d3d12 apps similar to d3d9 or opengl tracing is/was done without injector..
that could fix this two problems at one..
don't know if studying/reverse engineering dxcap.exe tool from Microsoft could help in knowing how this tool allows tracing D3D12 UWP apps also..
as said hope can help in some of this early steps..
Many thanks..

**Comment by jrfonseca** [comment]
> Can't trace UWP (universal Windows 10 platform/Winrt/Metro) apps: that was expected but posting here in case you know how to inject dxgitrace.dll to a UWP app...

Metro apps is a completely different ballgame.  It's not impossible to trace, and I do have some experimental code that enables tracing Metro apps but it's full of hacks (including undocumented interfaces), and extremely hard to use (lots of manual steps.)

I hope to polish it up at some point, but I rather to keep it separate from D3D12 support at this point.

> I first searched in injectee.cpp and added d3d12.dll thinking will fix it here below d3d11.dll:

That looks right.  I pushed it.

> with depends tool I examined both desktop&UWP app D3D12Fullscreen.exe and I see in executable d3d12.dll file d3d12createdevice isn't mentioned but there is a dependency function which is defined with ordinal 101 which for d3d12.dll is d3d12createdevice searching for strings in executable also shows d3d12createdevice isn't mentioned altough a reference to 

I believe the injection code does try to patch up imports by ordinal tool.  Maybe it's not working.

I'll look into it when I can.

But to unblock you, my recomendation is to: 1) only use desktop and ignore Metro 2) modify the sample app to do LoadLibraryA("d3d12.dll"); GetProcAddress("D3D12CreateDevice").  Or find a different sample app.

Also, D3D12SerializeRootSignature should be safe to ignore.   We could trace it too if those calls are useful.  But they are not required to replay.

> also thinking if would be posible to generate a d3d12.dll as way of tracing d3d12 apps similar to d3d9 or opengl tracing is/was done without injector..

No, I'm afraid that's impossible.  Because D3D12 interacts with DXGI and D3D11 which are in different DLLs, so if one doesn't intercept those calls, havoc will ensue as wrap / unwrapped COM objects get mixed up.

> don't know if studying/reverse engineering dxcap.exe tool from Microsoft could help in knowing how this tool allows tracing D3D12 UWP apps also..

I know how VS / dxcap works: the "legacy" capture (Windows 7) is essentially the same DLL injection mechanism, but on Windows 8 (or 8.1?) and later, the Direct3D runtimes (dxgi*.dll, d3d1*.dll) actually have a hooks which callback into Visual Studio capture DLLs.  This is what Microsoft calls "robust capture".  See https://msdn.microsoft.com/en-us/library/mt420574.aspx 

It wouldn't be impossible to try to leverage the same mechanism, so far I haven't seem any advantage.  I evaluated how dxcap works when I was struggling trying to trace Metro apps, but spite Metro apps being documented as supported in dxcap, unfortunately dxcap worked just as bad.  And since then I managed to trace Metro apps without relying on dxcap "robust capture".  So AFAICS, trying to mimick dxcap would force me to figure out a lot of undocumented Microsoft internals, and implement two different tracing methods (for Windows 7 and 8.1+) and give me nothing in return...

**Comment by oscarbg** [comment]
Thanks for your fast support and comments..
some thoughts:
*Dxcap works good for me.. in fact allowed to capture +2GB traces of Forza Apex and Quantum Break DX12 UWP apps like 1min of every game..
Perhaps it has improved in latest Win10 updates, mine is DXcap dated 13/11 from Win10 10586 build..
Also have latest VS 2015 U2 don't know if updated libraries it brings helps in robustness of my installed DXcap tool..
sad to hear mimicking dxcap  would require some hard work..
other good option would be if Microsoft documented the vsglog file format so ability to be parsed could be achieved..

*I will try to modify app using LoadLibraryA("d3d12.dll"); GetProcAddress("D3D12CreateDevice)..
but seems the problem I'm seeing with my D3D12 app is general as searching inside shipping D3D12 games like Forza Apex and Quantum Break binaries with depends tool I see same problem they reference 101 ordinal function in d3d12.dll like my used compiled sample..
Perhaps one reason injector is not working for me is I'm compiling with both apitrace and sample with Win10 SDK build 10586 not the RTM 10240.. perhaps d3d12 headers/libraries changed a little between these versions and even aren't ABI compatible as Forza and Quantum Break both explicitily say they require Win10 10586 or higher to run.. from what version of Win10 SDK headers you generated D3D12 support?

also one question: tried debugging Dllmain by placing breakpoint of dxgitrace.dll from dxgitrace project from VS by launching as debugger command "inject -D C:\code\apitrace-d3d12\bin64\Debug..\wrappers\Debug\dxgitrace.dll -- 
 C:\Users\oscarb\Downloads\DirectX-Graphics-Samples-1.0.160510.1\DirectX-Graphics-Samples-1.0.160510.1\Samples\UWP\D3D12Fullscreen\src\bin\x64\Release\D3D12Fullscreen.exe"
the app runs but the breakpoint DllMain is never reached so seems isn't possible to debug this library dll by launching the injector as specified, right? any help in debugging dxgitrace.dll correctly?

*Finally just seeing GpuPerfStudio 3.5 released today and says d3d12 capture plugin is open source as part of CodeXL on github.. perhaps code can reveal if some "magic" is applied to injecting D3D12CreateDevice for tracing DX12 apps..

**Comment by jrfonseca** [comment]
> other good option would be if Microsoft documented the vsglog file format so ability to be parsed could be achieved..

I don't see much point in actually parsing vslog.  apitrace/scripts/convert.py allows to convert pixrun/vsglog by tracing pix/dxcap executables.  The result should be the same.

Then again, if dxcap works well, then I'd just use it.  I maintain apitrace for Direct3D APIs because we (VMware) use it for test automation we always find a few things (e.g, unofficial extensions, or non MS APIs) that closed source tools don't support, where as we can extend apitrace to support them.   But I don't delude myself: it will never be as easy to use or as featureful as Visual Studio Graphics Debugger.   No matter how much producitivity I can extract from the little time I can afford to apitrace, I'll never be able to compete with the large teams behind such products.
 
> Finally just seeing GpuPerfStudio 3.5 released today and says d3d12 capture plugin is open source as part of CodeXL on github.. perhaps code can reveal if some "magic" is applied to injecting D3D12CreateDevice for tracing DX12 apps..

Most other tools use Mhook library for intercepting, whereas apitrace uses IAT patching.

I've been considering to switch to mhook, but it's not straightfoward port: with IAT I can selectively intercept the calls done from the application, but not the calls done by the wrapper, where as mhook modifies the actual code, so it applies to both to the app calls and the wrapper calls, so a lot of refactoring will be necessary to port.  

Anyway, these are orthogonal/tangention things.  Yes, we need to ensure we import by ordinal works, in order to fix D3D12CreateDevice (and I'll look into it ASAP), but supporting D3D12 requires much more than that.

**Comment by jrfonseca** [comment]
The ordinal import here worked just fine here when I tried.

I wonder if the problem here is that you're building 32 bits apitrace, while trying to trace a 64bits sample?  I actually had to modify the sample project to build on x86, because I only have a 32 bits Win 10 VM on my personal laptop.

There were a bunch of other problems. I pushed a few more fixes.  I'll have to stop now.

Don't forget to run DbgView to see apitrace messages.   I actually run https://github.com/jrfonseca/drmingw 's catchsegv tool to dump both the debug messages plus stack back traces when exceptions happen.

**Comment by SveSop** [comment]
Since there are some games that work with vkd3d and Wine for D3D12, having apitrace for devs to be able to replay crashes would be awesome.

**Comment by jrfonseca** [comment]
For reference of those interested there's now  
- https://github.com/apitrace/apitrace/pull/701
- https://github.com/Joshua-Ashton/apitrace/tree/d3d12 (see also https://blog.froggi.es/d3d12-apitrace-capture-and-replay/ )
Hopefully we'll D3D12 support merged in the near future.

**Comment by effolkronium** [comment]
Any plans for vulkan support ?

**Comment by jrfonseca** [comment]
For the record, I've asked Claude to rebase and cleanup the @misyltoad's https://github.com/misyltoad/apitrace/tree/d3d12 branch, add a test case, and support snapshots:

https://github.com/apitrace/apitrace/compare/d3d12-rebase
https://github.com/apitrace/apitrace-tests/compare/d3d12

I still need to find a bit more time to understand all changes and do more testing before merging, but it's a much smaller effort.
